test_application/src/main.rs

```diff
@@ -0,0 +1,272 @@
+use anyhow::{self, bail};
+use clap;
+use crossbeam_channel::{select, unbounded, Receiver, Sender};
+use rand::{rngs::OsRng, thread_rng};
+use sha2::{Digest, Sha256};
+use std::{collections::HashMap, ops::Sub, thread};
+
+use tracing::{debug, info, instrument, span, trace, Level};
+use tracing_subscriber::{self, EnvFilter};
+use tss_ecdsa::{
+    AuxInfoParticipant, AuxInfoPrivate, AuxInfoPublic, Identifier, KeySharePrivate, KeySharePublic,
+    KeygenParticipant, Message, Participant, ParticipantConfig, ParticipantIdentifier,
+    PresignInput, PresignParticipant, PresignRecord, ProtocolParticipant,
+};
+
+/// Messages from main thread to router thread.
+#[derive(Debug, Copy, Clone)]
+enum SubProtocol {
+    StartKeyGen,
+    StartAuxInfo,
+    StartPresign,
+    Sign,
+}
+
+/// Outgoing message from the Participant's POV.
+#[derive(Debug)]
+enum MessageFromWorker {
+    /// Worker made progress on sub-protocol, returning some messages.
+    FinishedRound(Vec<Message>),
+    /// Current sub-protocol ended.
+    SubProtocolEnded,
+}
+
+/// Maps Ids to channels for routing message to the correct participant based on
+/// [`ParticipantIdentifier`].
+type WorkerChannelsMessage = HashMap<ParticipantIdentifier, Sender<Message>>;
+/// Inform workers of the next [`SubProtocol`] to perform with the given
+/// [`Identifier`].
+type NextActionChannels = Vec<Sender<(SubProtocol, Identifier)>>;
+
+fn main() -> anyhow::Result<()> {
```

Comment:
I would also like your overall opinion on this example?

Reply:
I like it! I had to take a couple of passes to understand the channel setup and determine which functions are run centrally vs on threads since I'm not familiar with that kind of infrastructure. It would help me to have a little bit of documentation on the functions themselves to indicate role + which party (worker vs router) executes them. 

I like that it's small and that it doesn't get bogged down in handling all the potential failure cases. One potential gap is in handling caller-responsible error cases -- `CallerError` variant that means the router messed up somehow, and instead of having the entire worker bork in that case, it would be nice to keep it alive and have it inform the router of the mistake. It's not necessary here, of course, since the example should always work because all the workers are honest, but it would be helpful I think as a guide to users. But I'm glad there isn't a bunch of code about network failures and other types of mistakes.

Reply:
Makes sense, I would also like to see workers robust to failure and inform the main thread. We can fix this in the follow "networking" example.